### PR TITLE
[REVIEW] Add new `partition` API

### PR DESCRIPTION
**Pull request by jrhemstad**
Depends on https://github.com/rapidsai/cudf/pull/4471

As mentioned in https://github.com/rapidsai/cudf/issues/4427, `scatter_to_tables` is inherently problematic. The fundamental problem is that it has to materialize independent `cudf::table`s for each partition. It will be significantly better to instead return a single table and a set of offsets to the beginning of each partition (like how `hash_partition` and `round_robin_partition` work).

In that vein, this PR and adds `partition` to eventually replace `scatter_to_tables`. `partition` is a more generic form of `hash_partition` or `round_robin_partition` where the caller specifies for each row which partition it will belong to. 

Removing `scatter_to_tables` will be a follow up PR.

### CODE REVIEW

**Comment by harrism** [comment]
New features should go into 0.15 since 0.14 is a bug fix release.

File: cpp/tests/partitioning/partition_test.cpp [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    std::copy(result_offsets.begin(), result_offsets.end(), std::ostream_iterator<int32_t>(std::cout, ","));
+    std::cout << std::endl;
+
+    EXPECT_TRUE(std::equal(expected_offsets.begin(), expected_offsets.end(), result_offsets.begin()));
+
+    cudf::test::print(result.first->get_column(0));
+    std::cout << std::endl;
+
+    cudf::test::expect_columns_equal(expected_ints, result.first->get_column(0));
+}
```
Comment (harrism) [inline cpp/tests/partitioning/partition_test.cpp]:
```suggestion
}

```

**Comment by rjzamora** [comment]
@kkraus14 @jrhemstad - Looks like we need a few cython/python changes to use this in dask.  Is anyone already planning to tackle this, or should I get everything hooked up?

**Comment by kkraus14** [comment]
@galipremsagar is planning on plumbing this in once it lands. Thanks @rjzamora!

**Comment by jrhemstad** [comment]
> @galipremsagar is planning on plumbing this in once it lands. Thanks @rjzamora!

The eagle has landed. Does that mean the Python updates are in progress? 

@galipremsagar take note of https://github.com/rapidsai/cudf/issues/4607

The intention of `partition` is to take it's output an pass it into `split`, but due to the unfortunate API for `split`, you'll need to drop the first/last element of the partition offsets returned from `partition` in order to call `split`.

**Comment by rjzamora** [comment]
> @galipremsagar is planning on plumbing this in once it lands. Thanks @rjzamora!

Sounds good - I did start looking into this 10 minutes before your response, so feel free to dump on me if you get too busy @galipremsagar :)

**Comment by galipremsagar** [comment]
I'll pick it up as I'm free now. Will ping you @rjzamora incase I face any issue/need help.

@jrhemstad Sure, noted. Thanks for the heads-up. 👍

**Pull request merged**